pkg/controller/catalogsource/catalogsource_controller.go

```diff
@@ -0,0 +1,138 @@
+func Add(mgr manager.Manager, _ options.ControllerOptions) error {
+	return add(mgr, newReconciler(mgr))
+}
+
+// newReconciler returns a new reconcile.Reconciler
+func newReconciler(mgr manager.Manager) reconcile.Reconciler {
+	client := mgr.GetClient()
+	return &ReconcileCatalogSource{
+		client: client,
+	}
+}
+
+func add(mgr manager.Manager, r reconcile.Reconciler) error {
+
+	c, err := controller.New("catalogsource-controller", mgr, controller.Options{Reconciler: r})
+	if err != nil {
+		return err
+	}
+
+	_, defaultCatalogsources := defaults.GetGlobalDefinitions()
+	pred := predicate.Funcs{
+		CreateFunc: func(e event.CreateEvent) bool {
+			return false
+		},
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			if _, ok := defaultCatalogsources[e.MetaOld.GetName()]; ok {
+				return true
+			}
+			return false
+		},
+		DeleteFunc: func(e event.DeleteEvent) bool {
+			if _, ok := defaultCatalogsources[e.Meta.GetName()]; ok {
+				// If DeleteStateUnknown is true it implies that the Delete event was missed
+				// and we can ignore it.
+				if e.DeleteStateUnknown {
+					return false
+				}
+				return true
+			}
+			return false
+		},
+		GenericFunc: func(e event.GenericEvent) bool {
+			if _, ok := defaultCatalogsources[e.Meta.GetName()]; ok {
+				return true
+			}
+			return false
+		},
+	}
+
+	err = c.Watch(&source.Kind{Type: &olm.CatalogSource{}}, &handler.EnqueueRequestForObject{}, pred)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// blank assignment to verify that ReconcileOperatorHub implements reconcile.Reconciler
+var _ reconcile.Reconciler = &ReconcileCatalogSource{}
+
+// ReconcileCatalogSource reconciles a CatalogSource object
+type ReconcileCatalogSource struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+}
+
+func (r *ReconcileCatalogSource) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	log.Infof("Reconciling default CatalogSource %s", request.Name)
+
+	_, defaultCatalogsources := defaults.GetGlobalDefinitions()
+	defaultCatsrcDef := defaultCatalogsources[request.Name]
+	// Fetch the CatalogSource instance
+	instance := &olm.CatalogSource{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, instance)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			if !operatorhub.GetSingleton().Get()[defaultCatsrcDef.Name] {
+				createNewCatsrcInstance(r.client, defaultCatsrcDef)
+			}
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		return reconcile.Result{}, err
+	}
+
+	if instance.DeletionTimestamp != nil {
+		if !operatorhub.GetSingleton().Get()[defaultCatsrcDef.Name] {
+			createNewCatsrcInstance(r.client, defaultCatsrcDef)
+		}
+	}
+
+	if !defaults.AreCatsrcSpecsEqual(&defaultCatsrcDef.Spec, &instance.Spec) {
+		if err := r.client.Delete(context.TODO(), instance); err != nil {
```

Comment:
I'm not sure if we delete here that it will be gone by the time the `createNewCatsrcInstance()` is called in the next line -- does it make more sense to return here with a `Result{requeue: true}` here instead and have the above logic recreate it? 

Or alternatively, we could just update it directly if the specs don't match instead of delete/recreate. Using the SSA client would ensure this is a seamless process.

Reply:
Updated to return `Result{requeue:true, requeueAfter: time.Seconds * 5}` 

Deleting and recreating seems much cleaner, we used to do that for OperatorSources too

Reply:
LGTM